src/main/java/io/cryostat/discovery/KubeApiDiscovery.java

```diff
@@ -0,0 +1,623 @@
+                                        newNode.nodeType = nodeType.getKind();
+                                        newNode.children = new ArrayList<>();
+                                        newNode.target = null;
+                                        newNode.labels =
+                                                kubeObj != null
+                                                        ? kubeObj.getMetadata().getLabels()
+                                                        : new HashMap<>();
+                                        // Add namespace to label to retrieve node later
+                                        newNode.labels.put(
+                                                DISCOVERY_NAMESPACE_LABEL_KEY, namespace);
+                                        return newNode;
+                                    });
+            return Pair.of(kubeObj, node);
+        }
+    }
+
+    @ApplicationScoped
+    static final class KubeConfig {
+        public static final String KUBERNETES_NAMESPACE_PATH =
+                "/var/run/secrets/kubernetes.io/serviceaccount/namespace";
+
+        private static final String OWN_NAMESPACE = ".";
+
+        @Inject Logger logger;
+        @Inject FileSystem fs;
+
+        @ConfigProperty(name = "cryostat.discovery.k8s.namespaces")
+        Optional<List<String>> watchNamespaces;
+
+        @ConfigProperty(name = "kubernetes.service.host")
+        Optional<String> serviceHost;
+
+        private KubernetesClient kubeClient;
+
+        Collection<String> getWatchNamespaces() {
+            return watchNamespaces.orElse(List.of()).stream()
+                    .map(
+                            n -> {
+                                if (OWN_NAMESPACE.equals(n)) {
+                                    return getOwnNamespace();
+                                }
+                                return n;
+                            })
+                    .filter((n) -> !n.isBlank())
+                    .collect(Collectors.toSet());
+        }
+
+        String getOwnNamespace() {
+            try {
+                return fs.readString(Path.of(KUBERNETES_NAMESPACE_PATH));
+            } catch (Exception e) {
+                logger.trace(e);
+                return null;
+            }
+        }
+
+        boolean kubeApiAvailable() {
+            return serviceHost.isPresent();
+        }
+
+        KubernetesClient kubeClient() {
+            if (kubeClient == null) {
+                kubeClient =
+                        new KubernetesClientBuilder()
+                                .withTaskExecutor(Infrastructure.getDefaultWorkerPool())
+                                .build();
+            }
+            return kubeClient;
+        }
+    }
+
+    private final class EndpointsHandler implements ResourceEventHandler<Endpoints> {
+        @Override
+        public void onAdd(Endpoints endpoints) {
+            logger.infov(
```

Comment:
#349 - I think the log level for these kinds of events should probably be `debug`